### ISSUE TITLE
Load Flux base model from quantized?

### ISSUE THREAD

**Issue by IdiotSandwichTheThird** [bghira/SimpleTuner#681]
Hi, I've got 32gb of ram. Every time I start up a lora training run, it takes ~17 minutes to finish quantizing Flux and start training due to using another ~40gb of swap. I think it would be ultra cool if we could avoid that by loading a pre-quantized model.

**Comment by bghira** [comment]
unfortunately it doesn't really work. it's still quite slow to load the pre-quantised weights.

makes me feel grateful for the Apple Silicon that quantises in about 20 seconds

**Comment by bghira** [comment]
you might just need more system memory, or a faster swap device

**Comment by TheDuckingDuck** [comment]
> unfortunately it doesn't really work. it's still quite slow to load the pre-quantised weights.
> 
> makes me feel grateful for the Apple Silicon that quantises in about 20 seconds

Kinda curious how comfy can load the whole model in fp8 so quickly in comparison. I get that's inference only, but does it make that much of a difference when you're only training lora?

**Comment by bghira** [comment]
i'm talking about when testing the trainer, it quantises that quickly

**Comment by TheDuckingDuck** [comment]
(I'm OP btw)
Okay, so the main reason for the request was of course the slow loading and ram usage. Apparently a big part of it was that I wanted to use adafactor, so I had to set base_model_default_dtype to fp32. This casts the entire base model to fp32 before quantizing, which seems like an entirely unnecessary step to me, as long as we keep the lora weights in fp32. So I edited the train.py to not cast the base model to fp32 and just quantize in bf16, and as far as I can tell things work fine.

Training starts within 5 minutes instead of 17 now.

**Comment by bghira** [comment]
you can't, though - adafactor isn't designed to handle bf16 weights or rounding bias issues. this is only handled by adamw_bf16. why use adafactor?

**Comment by TheDuckingDuck** [comment]
> you can't, though - adafactor isn't designed to handle bf16 weights or rounding bias issues. this is only handled by adamw_bf16. why use adafactor?

Excuse me if I'm wrong, but the actual weight we're training (the lora weights) are kept in fp32, I'm just avoiding casting the entire model weights unnecessarily, since those are quantized to fp8 anyways, almost immediately.

Unless there's a difference between fp8(bf16) and fp8(fp32)? I kinda assumed it all gets cast to bf16 anyways after quantizing.

Talking about changing only this specific line to always bf16.
https://github.com/bghira/SimpleTuner/blob/main/train.py#L858

**Comment by bghira** [comment]
it's fine - it's easy to be confused by the order of operations there. but on L1602 we move the transformer weights incl the adapter to the selected weight_dtype. they're always the same precision unless you're specifically using `--base_model_default_dtype` to switch them, but the main precision of bf16 refers to both by default.

**Comment by bghira** [comment]
low-memory systems probably are best served with adamw_bf16 since it loads everything in bf16 from the very start and keeps the weights there the whole time. adafactor requires fp32 parameters since we've not modified it to support bf16.

**Comment by bghira** [comment]
that's really a question for Comfy.

**Comment by Nuclear6** [comment]
@bghira https://huggingface.co/Comfy-Org/flux1-dev/blob/main/flux1-dev-fp8.safetensors
Sorry, I still don't understand why others have already quantized the fp8 model, but we can't train directly. Why do we have to re-quantize it before we can start training?

**Comment by bghira** [comment]
because it isn't the same type of quantisation - training requires a very specific configuration.

**Comment by kaibioinfo** [comment]
@sneccc it's a problem in ComfyUI and I'm not even sure if it's solvable. The way Loras work in ComfyUI is that they are merged into the base model (instead of running them online). This means that if the lora changes the model only a tiny little bit, this change might got lost due to the quantisation/low precision. Maybe it helps if one would temporarily upcast the weights before applying the lora, I could check that.

**Comment by bghira** [comment]
funny enough it looks like the x-flux trainer just loads the fp8 weights to cast them to fp32 anyway

**Comment by mhirki** [comment]
@sneccc Just out of curiosity, is ComfyUI loading the model in lowvram mode, i.e. does it say "loaded in lowvram mode" in the console?

**Comment by bghira** [comment]
i've been testing the x-flux settings:

- PURE_BF16=true
- MIXED_PRECISION=bf16
- LR_SCHEDULER=constant
- LEARNING_RATE=1e-5
- TRAIN_BATCH_SIZE=1
- GRADIENT_ACCUMULATION_STEPS=2
- TRAINER_EXTRA_ARGS="--lora_rank=16 --lora_alpha=16 --max_grad_norm=1.0 --gradient_precision=fp32 --base_model_default_dtype=bf16 --lora_init_type=loftq --flux_lora_target=mmdit  --user_prompt_library=user_prompt_library.json --compress_disk_cache"
- OPTIMIZER="adamw_bf16"
- resolution: 512px

It's going 1 second per step and uses 40G VRAM.

It's starting to adapt the dataset, though it first focuses on all of the artifacts from the cell phone pictures:

![image](https://github.com/user-attachments/assets/07342dcf-3598-467c-b9ab-8114e2db84b0)

**Comment by bghira** [comment]
yucky 🤣 no probably not

**Comment by mhirki** [comment]
> > @sneccc Just out of curiosity, is ComfyUI loading the model in lowvram mode, i.e. does it say "loaded in lowvram mode" in the console?
> 
> idk what u mean low vram mode, i have 24gb vram i think its managed automatically, unless u mean fp8

Yes, ComfyUI automatically offloads parts of the model to system RAM. ComfyUI currently has a bug where it doesn't apply LoRAs properly when there isn't much VRAM available. But since you have 24 GBs of VRAM, it probably doesn't affect you much. Although bear in mind lowvram mode is triggered even with a 24 GB card when loading flux in "default" precision.
